[IV-25-26] Objetivo 2

chelunike opened the pull request:
Issue:  #7

Sobre la estructura del repositorio

* [x] ¿He seguido las mejores prácticas en el nombre de las clases y ficheros y
   disposición de los mismos?
* [x] ¿He mirado tanto reglas generales (para cualquier repo) como reglas
específicas para el lenguaje con el que se está trabajando?

Sobre el planteamiento

* [x] ¿Se han planteado una serie de issues, asignados a las historias de
usuario pertinentes, que clarifiquen qué es lo que necesitan los usuarios e
identifiquen las diferentes partes del problema?
* [x] ¿Los issues representan un problema, y no una tarea?

Sobre el análisis del problema

* [x] ¿Se ha documentado qué análisis se ha hecho sobre el dominio para decir lo
   que se ha creado?
* [x] ¿Se ha documentado por qué se ha elegido que lo creado sea un objeto valor,
   una entidad o un agregado?
* [x] ¿He propuesto un producto mínimamente viable, que en muchos casos será un solo
   objeto valor que no dependa de ningún otro (y que sea la base de muchos
   otros)?.

Sobre el código

* [x] ¿Todos los mensajes de commit explican el cambio, y no se
      limitan a repetir el nombre del fichero que se ha cambiado?
* [x] ¿Los mensajes de commit siguen el formato estándar y buenas prácticas?
* [x] ¿Se ha hecho una revisión real del código para comprobar que todos
   los atributos y funciones creadas están respaldadas por una HU?
* [x] ¿Todos los issues creados están relacionados con a una HU?
* [x] ¿Ha asignado el propietario/la propietaria del repositorio todos los
issues planteados al *milestone* correspondiente?
* [x] ¿Todos los cambios en el código están asignados a un issue al que se
      referencia en un commit?
* [x] ¿Se ha asignado al mismo milestone el PR que se ha hecho?
* [x] ¿Es el milestone sobre el que estoy trabajando un producto
   mínimamente viable? ¿O tengo que solicitar al product manager que precise de
   qué producto se trata?
* [x] ¿Se ha comprobado que el código entregado sea sintácticamente
correcto?

Lista de comprobación de @gabrielherreraloz 

* [x] ¿Los issues que ha planteado la persona son efectivamente problemas y evidencian comprensión de las historias de usuario? Si no es así, es mejor que se le transmita antes de que empiece a crear cualquier código erróneo. 
* [x] ¿Has comprobado que efectivamente el compañero/a ha seguido *su* lista de comprobación? 
* [x] ¿El código creado es una modelización válida de las historias de usuario creadas por ti, y te sirve como base para programar la lógica de negocio? 
* [x] ¿El código creado sigue las mejores prácticas para el lenguaje de programación que hayáis consensuado, tanto en identificadores, como nombres de ficheros, como disposición de los diferentes ficheros en el repositorio? Por ejemplo, ¿todos los commits se refieren a un issue y aportan código para resolver el problema planteado en el mismo? 
* [x] ¿He comprobado que el código compile, o que quien lo ha programado ha lo ha hecho y documentado en el repositorio? 
* [x] ¿He añadido el PR y todos los issues que se resuelvan en él al milestone correspondiente, que será el primero? 
* [x] ¿Cubre totalmente el milestone 0, o hará falta modelización adicional en el objetivo 4 para poder programar la lógica de negocio? 
* [x] En resumen, ¿considero que con lo hecho puedo ponerme a trabajar directamente en el siguiente milestone, que incluye la lógica de negocio?

JJ commented:
Revisión @jacarmona364

chelunike commented:
Revisión @gabrielherreraloz

gabrielherreraloz commented on internal/persona/model.go:
```diff
@@ -0,0 +1,8 @@
+    ID   string `json:"id"`
+    Nombre string `json:"nombre"`
+    Edad int `json:"edad"`
+    PesoMochila float32 `json:"pesoMochila"`
```
Lo que hablamos, no tiene sentido guardar el valor del peso de la mochila

chelunike replied:
He eliminado el archivo, y pese a ello, no ha cerrado la review. Procedo a cerrarla a mano.

gabrielherreraloz reviewed (commented):
No se si es adecuado hacer la clase Persona en este objetivo puesto que no es información que tengamos disponible ahora, la tendría que dar el usuario, por ahora solo tenemos lo que podamos encontrar en la web, tenemos los distintos tipos de medio de transporte, las líneas y las paradas.

chelunike committed:
Eliminar clase innecesaria de Persona
No es necesaria para #10, porque según #3 la persona elige la mejor opción por si misma y no requiere de representarla.

chelunike committed:
Ajustando representación del medio de transporte a los datos obtenidos en #10

chelunike committed:
Añadida representación de la parada del medio de transporte para #16
He añadido los datos que representan minimamente la parada para representar la ruta.

chelunike committed:
Cambiando nombre del modelo al correcto para #16

chelunike committed:
Añadida lista de paradas close #16
Necesimatamos representar la lista de paradas para poder representar la ruta que necesitamos obtener en #16

gabrielherreraloz reviewed (commented):
Lo veo bastante bien, lo único que incluiría más es poner en parada, el numero de líneas de esa parada, es decir, los transportes que pasan por ella. Por lo demás lo veo bastante correcto.

chelunike committed:
Añadida lista de correspondencias de lineas con la parada.
Tal y como ha indicado @gabrielherreraloz es necesario almacenar esa lista de referencias al código de la línea, que suele ser un número entero, pero hay muchos casos en los que varia y contiene letras. E incluso hay lineas que se referencian con solo letras.

gabrielherreraloz commented on internal/parada/model.go:
```diff
@@ -0,0 +1,8 @@
+    nombre string `json:"nombre"`
+    latitud float64 `json:"latitud"`
+    longitud string `json:"longitud"`
+    lineasCorrespondencia []string `json:"lineasCorrespondencia"`
```
Creo que aquí tendría mas sentido guardar los objetos de la clase Transporte en lugar de un array de strings, y de forma que se pudieran identificar por alguna clave para tener una buena relación Parada-Transporte

gabrielherreraloz commented on internal/transporte/model.go:
```diff
@@ -0,0 +1,10 @@
+    tipoMedio string `json:"tipoMedio"`
+    nombre string `json:"nombre"`
+    numLinea string `json:"numLinea"`
+    paradas []Parada `json:"paradas"`
```
Y aquí lo mismo que en el otro array

JJ replied:
Una línea incluye paradas y las paradas incluyen líneas. ¿Cuál es la única fuente de verdad para paradas y líneas?

chelunike committed:
Cambio de las lineas de correspondencia a Map
Para poder guardar todas las lineas de correspondencia con sus datos y organizarlas por su código correspondiente.

chelunike commented:
Revisión @JJ

JJ commented on internal/parada/model.go:
```diff
@@ -0,0 +1,10 @@
+    nombre string `json:"nombre"`
+    latitud float64 `json:"latitud"`
+    longitud string `json:"longitud"`
+    lineasCorrespondencia map[string]Trasnporte `json:"lineasCorrespondencia"`
```
Por favor, lee bien ese nombre

chelunike replied:
No termino de ver el error evidente, puedes especificar un poco más cual es "ese nombre". :sweat_smile:

JJ replied:
No, no puedo. Por favor, mira bien. O trata de compilar, que es lo que debías haber hecho desde el principio

JJ commented on internal/transporte/model.go:
```diff
@@ -0,0 +1,10 @@
+import "internal/parada"
+
+type Transporte struct {
+    tipoMedio string `json:"tipoMedio"`
```
El que lo leas directamente del JSON no es parte de este hito ni forma parte del dominio del problema.

JJ commented on iv.yaml:
```diff
@@ -0,0 +1,2 @@
+lenguaje: go
+entidad: internal/persona/model.go
```
Este path no existe.

JJ reviewed (commented):
Aparte, no entiendo qué ventaja puede tener un directorio con un solo fichero frente a un fichero.

JJ commented on internal/parada/model.go:
```diff
@@ -0,0 +1,10 @@
+
+type Parada struct {
+    nombre string `json:"nombre"`
+    latitud float64 `json:"latitud"`
```
¿Un PMV requiere la latitud y la longitud? ¿La historia de usuario lo pide?

chelunike replied:
Si, porque el usuario ubica las paradas más cercanas a su persona, y para ello es necesario ubicar la parada.

JJ replied:
¿Cómo vas a hacer eso sin un API?
En todo caso, ubicar paradas ¿correspondería al dominio de la solución o del problema?
Recueda que si lo único que estás haciendo es cambiar de nombre a la información que tienes y más o menos organizándola, no estás haciendo en realidad un modelo.

JJ commented on internal/transporte/model.go:
```diff
@@ -0,0 +1,10 @@
+
+import "internal/parada"
+
+type Transporte struct {
```
Si esto es una línea de correspondencia, ¿por qué se llama transporte?

chelunike replied:
Esta actualizado para llamarlo todo como transporte

JJ replied:
Repito la pregunta: Si esto es una línea de correspondencia, ¿por qué se llama transporte?

chelunike replied:
De acuerdo tratamos con la nomenclatura de Linea de correspondencia y Solo trabajamos con lineas

JJ commented on internal/transporte/model.go:
```diff
@@ -0,0 +1,10 @@
+type Transporte struct {
+    tipoMedio string `json:"tipoMedio"`
+    nombre string `json:"nombre"`
+    numLinea string `json:"numLinea"`
```
¿Por qué hace falta el nombre *y* el número de línea?

chelunike replied:
Porque el usuario necesita esa información para reconocer al transporte.

JJ replied:
Una vez más, lo que necesita el usuario ¿es parte del dominio del problema o de la solución?
Recuerda de nuevo que si estás poniendo en una estructura de datos lo que está en la página web, *no estás haciendo un modelo*

JJ reviewed (commented):
Comentarios adicionales.
No creo que puedas hacerlo esta mañana, así que tranquilo, hazlo ya para la convocatoria extraordinaria.

chelunike committed:
Actualizando path del modelo
Ha cambiado el path del modelo y no se ha actualizado

chelunike committed:
Eliminando referencias del json
No forman parte del dominio del problema y no se tratan para este objetivo

chelunike committed:
Cambiado nombre de lineasCorrespondencia a correspondenciaTransportes
Para evitar errores de nomenclatura y hablar siempre de transportes y paradas.

chelunike committed:
Asegurando una única fuente de verdad con referencias
Con las referencias nos aseguramos de que solo disponemos de un objeto como referencia de verdad.

chelunike committed:
Eliminando latitud y longitud
No forman parte del PMV ni de este objetivo

chelunike committed:
Cambiando nomenclatura a Linea y correspondenciaLineas
Para solo trabajar con los datos del dominio del problema, hablado con @gabrielherreraloz que serán los del consorcio de transportes.

JJ reviewed (commented):
Sigue sin haber una única fuente de verdad ni para Paradas ni para líneas.

vpedrosa commented on internal/parada/model.go:
```diff
@@ -0,0 +1,8 @@
+package parada
+
+import "internal/transporte"
```
Referencia circular aquí, revisa errores al compilar.

vpedrosa commented on internal/transporte/model.go:
```diff
@@ -0,0 +1,10 @@
+package transporte
+
+import "internal/parada"
```
Referencia circular aquí, revisa errores al compilar.

vpedrosa reviewed (commented):
Hola @chelunike, te dejo una revisión en base a mis errores. Los issues #10 y #16 están planteados como tareas de obtención de datos ("Necesitamos obtener...") en lugar de problemas del dominio derivados de la HU. Además, no se han creado errores específicos del dominio, que según el objetivo son parte del modelo ("Gestión de Excepciones: Diseñar jerarquía de excepciones coherente"). Por último, los commits no documentan por qué Linea y Parada son entidades y no objetos valor. Te dejo comentarios específicos sobre los problemas que te ha indicado ya @JJ .

chelunike committed:
Unificando modelos en un mismo fichero
Tal y como ha mencionado @jj en un comentario de #8 NO es necesario separarlo en dos entidades.

chelunike commented:
@JJ En GO no existe "readonlly", ni constantes en estructuras de datos, solo existen constantes en tiempo de compilación. Puedo reflejar esa única fuente de verdad a traves de lógica de negocio, definiendo funciones que no permitan la duplicidad de datos y solo almacenen una única versión de esa entidad.

JJ commented:
La inmutabilidad sólo tiene que ver parcialmente con la SSOT. No puedes definir funciones en esta fase, no está en el hito.

chelunike commented:
Si no puede definir inmutabilidad por el lenguaje, al menos en la parte que corresponde al objetivo, si ya está definido las relaciones a traves de referencias. Y no está permitida la escritura al objeto, al ser privado esos atributos, entonces ya definiriamos la SSOT, dentro de nuestras posibilidades.

JJ commented:
> Si no puede definir inmutabilidad por el lenguaje, al menos en la parte que corresponde al objetivo, si ya está definido las relaciones a traves de referencias. Y no está permitida la escritura al objeto, al ser privado esos atributos, entonces ya definiriamos la SSOT, dentro de nuestras posibilidades.

Insisto, la inmutabilidad no tiene nada que ver son que sean o no SSOT. La inmutabilidad es para los objetos valor, una SSOT es generalmente una entidad. Pero en fin, tú haz la propuesta en el PR y ya lo vemos viendo.

vpedrosa commented:
> @JJ En GO no existe "readonlly", ni constantes en estructuras de datos, solo existen constantes en tiempo de compilación. Puedo reflejar esa única fuente de verdad a traves de lógica de negocio, definiendo funciones que no permitan la duplicidad de datos y solo almacenen una única versión de esa entidad.

Creo que estás enfocándolo mal. "Única fuente de verdad" no es un problema de inmutabilidad, es un problema de arquitectura. Única fuente de verdad significa, a mi parecer, que cada entidad (Parada, Línea) debe existir en un único lugar, y el resto del código debe referenciarla, no copiarla.

¿Has valorado crear repositorios/almacenes centralizados de entidades y que el resto del sistema las referencie por ID o con un puntero?

chelunike committed:
Añadiendo ID a la Linea
Para asegurarnos una única fuente de verdad, podemos añadir un número de ID unico a cada línea y nos aseguramos que la información sale de la misma fuente.

gabrielherreraloz commented on internal/transporte/model.go:
```diff
@@ -0,0 +1,18 @@
+
+type Linea struct {
+    // Identifica a la linea
+    id uint
```
Vuelvo a lo mismo que te comentó @vpedrosa , la cosa no es que el id vaya dentro del struct, deberías enfocarlo de forma que, por ejemplo, todo se guarde en un único contenedor, cada elemento identificado con un id.

gabrielherreraloz commented on internal/transporte/model.go:
```diff
@@ -0,0 +1,18 @@
+type Linea struct {
+    // Identifica a la linea
+    id uint
+    tipoMedio string
```
Aquí, en vez de ser un string, podrías crear un tipo para que solo pueda asignarse un tipo de los definidos ya que no hay muchos y así queda como algo más fijo.

gabrielherreraloz replied:
@chelunike no se si esto lo viste

gabrielherreraloz reviewed (commented):
Te he comentado un par de cosas importantes, la segunda no tanto, pero la primera es clave.

chelunike committed:
Eliminando ID de la Linea y tratandola desde fuera.
Para asegurarnos de trabajar con la misma linea y parada, podemos usar un map en las que se identifiquen por ID.

chelunike committed:
Añadido tipo de medio
Añadido iota (Un tipo especifco) para clasificar el tipo de medio de transporte. Tal y como hemos hablado en #8

chelunike commented:
Revisión @vpedrosa

oscar0310 reviewed (commented):
Buenas @chelunike , a la hora de hacer los commits hay muchos que no tienen un issue referenciado , esto provoca que no se entienda bien porqué haces el código, te recomiendo de que sigas una correcta metodología de trabajo y referencies al issue que te estas refiriendo con el número del issue en el título del commit.

gabrielherreraloz commented on internal/transporte/model.go:
```diff
@@ -0,0 +1,26 @@
+}
+
+// Almacenamos las lineas a partir de un ID único.
+lineas = map[uint]Linea
```
Esto también da error de sintaxis

vpedrosa replied:
Error de sintaxis. No debes subir código sin compilar, te ha pasado un par de veces.

gabrielherreraloz commented on internal/transporte/model.go:
```diff
@@ -0,0 +1,26 @@
+
+type TipoMedio int
+const (
+    AUTOBUS = TipoMedio iota
```
```suggestion
    AUTOBUS TipoMedio = iota
```

si no, da error de sintaxis

vpedrosa commented on internal/transporte/model.go:
```diff
@@ -0,0 +1,26 @@
+    tipoMedio TipoMedio
+    nombre string
+    numLinea string
+    paradas []*Parada
```
Al no tener un `map[uint]*Parada` no garantizas que cada parada sólo exista una vez.

vpedrosa reviewed (commented):
El código sigue sin única fuente de verdad: falta un almacén centralizado para Paradas, el almacén de Lineas tiene errores (falta var y debería usar punteros), y Linea contiene copias  directas de Paradas en lugar de referencias por ID. Necesitas crear dos almacenes globales (var paradas y var lineas con punteros) y hacer que ambas entidades se referencien  consistentemente por IDs, no guardando copias de objetos completos.

chelunike committed:
Solucionando errores de compilación como se indica en #8
Y añadiendo una correcta lista de paradas y lineas tal y como se indica en los comentarios.
Tratando de completar el objetivo 2 de acuerdo al issue #7 y #10

chelunike commented:
Solicito revisión, por favor @vpedrosa  @gabrielherreraloz

gabrielherreraloz commented on internal/transporte/model.go:
```diff
@@ -0,0 +1,27 @@
+    tipoMedio TipoMedio
+    nombre string
+    numLinea string
+    paradas [uint]*Parada
```
Esto no compila, no es una asignación correcta de tamaño. Dentro de los corchetes debe estar vacío o una constante.

gabrielherreraloz reviewed (commented):
Sigue sin compilar

chelunike committed:
Corrigiendo error de compilación en Linea tal y como se menciona en #8

chelunike commented:
Revisión @JJ por favor :)

JJ commented on internal/transporte/model.go:
```diff
@@ -0,0 +1,27 @@
+}
+
+// Almacenamos las lineas a partir de un ID único.
+var paradas = map[uint]*Parada{}
```
¿De veras crees que una variable global es la mejor forma de gestionar esto?
¿No has pensado que puede haber varias personas o varias áreas diferentes con las que trabajar?

gabrielherreraloz replied:
Estoy de acuerdo, podrías seguir la estructura como las que has hecho arriba y guardarlo todo en un tipo de almacén...

JJ reviewed (commented):
El concepto de única fuente de verdad se refiere a un tipo de estructura de datos. Y las variables globales son *siempre* una mala idea.

chelunike committed:
Añadida entidad Almacen para aislar la fuente de los datos
Tal y como hemos comentado en #8

JJ reviewed (approved):
👍

chelunike committed:
Merge remote-tracking branch 'gabri/main' into Objetivo-2
Para resolver los conflictos de #8
